common/lib/client/channel.ts

```diff
@@ -0,0 +1,178 @@
+import * as Utils from '../util/utils';
+import EventEmitter from '../util/eventemitter';
+import Logger from '../util/logger';
+import Presence from './presence';
+import Crypto from 'platform-crypto';
+import Message, { CipherOptions } from '../types/message';
+import ErrorInfo from '../types/errorinfo';
+import PaginatedResource, { PaginatedResult } from './paginatedresource';
+import Http from 'platform-http';
+import Resource from './resource';
+import { ChannelOptions } from '../../types/channel';
+import { PaginatedResultCallback } from '../../types/utils';
+
+// TODO: Replace these when Realtime and Rest are in TypeScript
+type Realtime = any;
+type Rest = any;
+
+interface RestHistoryParams {
+	start?: number;
+	end?: number;
+	direction?: string;
+	limit?: number;
+}
+
+function noop() {}
+const MSG_ID_ENTROPY_BYTES = 9;
+
+function allEmptyIds(messages: Array<Message>) {
+	return Utils.arrEvery(messages, function(message: Message) {
+		return !message.id;
+	});
+}
+
+function normaliseChannelOptions(options?: ChannelOptions) {
+	const channelOptions = options || {};
+	if(channelOptions.cipher) {
+		if(!Crypto) throw new Error('Encryption not enabled; use ably.encryption.js instead');
+		const cipher = Crypto.getCipher(channelOptions.cipher);
+		channelOptions.cipher = cipher.cipherParams;
+		channelOptions.channelCipher = cipher.cipher;
+	} else if('cipher' in channelOptions) {
+		/* Don't deactivate an existing cipher unless options
+			* has a 'cipher' key that's falsey */
```

Comment:
Hmm. I wasn't able to associate this comment and the lines below 🤔

Reply:
So the first branch of the if statement is `if (channelOptions.cipher)` which checks if `channelOptions.cipher` is truthy. The `else if` branch merely checks if `channelOptions.cipher` exists at all, so it will only trigger if the `cipher` is falsey. So essentially this branch will 'deactivate' the cipher if `channelOptions.cipher` is falsey. Perhaps it would be equivalent to say 'Deactivate the existing cipher if the 'cipher' key is falsey'.

Reply:
Resolving this conversation on @ikbalkaya's behalf.